### PR TITLE
[FLINK-25228][table] Introduce flink-table-test-utils

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

Goal of this new module is to provide a easy to use module for testing formats, connectors and end users.

The module brings in transitively all the dependencies you need to run a table test and includes a couple of test utilities.

## Brief change log

* Introduce `flink-table-test-utils`, shipping the existing table assertions and `TableFactoryHarness`. Both assertions and `TableFactoryHarness` are marked as `@Experimental`
* Expand the existing table assertions
* Show the usage of the assertions

## Verifying this change

The newly introduced assertions are tested in a new test, showing how to use them.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-test-utils/pom.xml [hunk 1]
```diff
@@ -107,6 +107,11 @@ under the License.
 									<includes>
 										<include>org/apache/flink/table/test/**</include>
 									</includes>
+									<excludes>
+										<!-- These are manually overridden by this package -->
```
Comment (twalthr) [inline flink-table/flink-table-test-utils/pom.xml:111]: Same question as above: Why not just move it to `utils`?

File: flink-table/flink-table-test-utils/src/test/java/TableAssertionTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import static org.apache.flink.table.api.DataTypes.ARRAY;
+import static org.apache.flink.table.api.DataTypes.BOOLEAN;
+import static org.apache.flink.table.api.DataTypes.FIELD;
+import static org.apache.flink.table.api.DataTypes.INT;
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.test.TableAssertions.assertThat;
+import static org.apache.flink.table.test.TableAssertions.assertThatRows;
+
+class TableAssertionTest {
```
Comment (twalthr) [inline flink-table/flink-table-test-utils/src/test/java/TableAssertionTest.java:39]: List in the java docs with classes it tests

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @slinkydeveloper. I totally support this PR but I would like to remove confusion in the code base and duplicate code. Let's move classes from `table-common` and other locations to the new test utils package. If it is only about `DataStructureConverter` we can also perform a class lookup whether `flink-table-runtime` is present or not. That should be fine for test utils.

**Comment by slinkydeveloper** [comment]
@twalthr the reason for placing the classes in two places is to continue to use the asserts in table-common as well.

> If it is only about DataStructureConverter we can also perform a class lookup whether flink-table-runtime is present or not

That's something I can do with some reflections. This way, we can still use the assertions within table-common (without some methods that requires the runtime dep), and then we don't have duplicated code. But we'll still need the shading. Is it ok for you?

File: flink-table/flink-table-test-utils/pom.xml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+			<version>${project.version}</version>
+			<classifier>tests</classifier>
+			<type>test-jar</type>
+			<scope>compile</scope>
+			<!-- We don't want table-common to be used as transitive dependency -->
+			<optional>true</optional>
+		</dependency>
+
+		<!-- Required for the assertions -->
+		<dependency>
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-test-utils/pom.xml:82]: `flink-test-utils` pulls `flink-test-utils-junit`, but we need to redeclare it here as the parent pom somewhere declares it with scope test, so we need to override the scope here. Removing `flink-test-utils-junit` returns this:

```
[INFO] org.apache.flink:flink-table-test-utils:jar:1.15-SNAPSHOT
[INFO] +- org.apache.flink:flink-test-utils:jar:1.15-SNAPSHOT:compile
[INFO] |  +- org.apache.flink:flink-test-utils-junit:jar:1.15-SNAPSHOT:test
```

File: flink-table/flink-table-test-utils/pom.xml [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-table</artifactId>
+		<version>1.15-SNAPSHOT</version>
+		<relativePath>..</relativePath>
+	</parent>
+
+	<artifactId>flink-table-test-utils</artifactId>
+	<name>Flink : Table : Test Utils</name>
+	<description>
+		This module contains test utilities for the Table API/SQL ecosystem.
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-test-utils/pom.xml:33]: I doubled down, check now :)

**Comment by slinkydeveloper** [comment]
@twalthr ready for another pass

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java [hunk 5]
```diff
@@ -95,7 +95,11 @@
  *    }
  * }
  * }</pre>
+ *
+ * @deprecated If you're using this test utility, then move the test class in flink-table-tests and
+ *     use the TableFactoryHarness from table-test-utils
  */
+@Deprecated
 public class TableFactoryHarness {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TableFactoryHarness.java:103]: We still have duplicate classes. We should port the existing tests and remove this class.

File: flink-table/flink-table-test-utils/src/main/java/org/apache/flink/table/test/factories/TableFactoryHarness.java [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+ *        }
+ *    }
+ * }
+ * }</pre>
+ */
+@Experimental
+public class TableFactoryHarness {
+
+    /** Factory identifier for {@link Factory}. */
+    public static final String IDENTIFIER = "harness";
```
Comment (twalthr) [inline flink-table/flink-table-test-utils/src/main/java/org/apache/flink/table/test/factories/TableFactoryHarness.java:104]: This factory is not registered in META-INF of this module.

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @slinkydeveloper. We still need to remove the duplicate `TableFactoryHarness`.

**Comment by slinkydeveloper** [comment]
@twalthr as discussed, I've removed the table factory harness

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**